PyDynamic/uncertainty/propagate_DFT.py

```diff
@@ -1,776 +1,908 @@
+        F: np.ndarray of shape (2M,)
+            real and imaginary parts of the second factor
+        UY: np.ndarray either shape (2M,) or shape (2M,2M)
+            covariance matrix or squared uncertainty associated with Y
+        UF: np.ndarray of shape (2M,2M)
+            covariance matrix associated with F (optional), default is None
+
+    Returns
+    -------
+        YF: np.ndarray of shape (2M,)
+            the product of Y and F
+        UYF: np.ndarray of shape (2M,2M)
+            the uncertainty associated with YF
+    """
+
+    assert (len(Y) == len(F))
+
+    def calcU(A, UB):
+        # uncertainty propagation for A*B with B uncertain (helper function)
+        n = len(A)
+        RA = A[:n // 2]
+        IA = A[n // 2:]
+        if isinstance(UB,
+                      float):  # simpler calculation if only single uncertainty
+            uRR = RA * UB * RA + IA * UB * IA
+            uRI = RA * UB * IA - IA * UB * RA
+            uII = IA * UB * IA + RA * UB * RA
+        elif len(UB.shape) == 1:  # simpler calculation if no correlation
+            UBRR = UB[:n // 2]
+            UBII = UB[n // 2:]
+            uRR = RA * UBRR * RA + IA * UBII * IA
+            uRI = RA * UBRR * IA - IA * UBII * RA
+            uII = IA * UBRR * IA + RA * UBII * RA
+        else:  # full calculation because of full input covariance
+            UBRR = UB[:n // 2, :n // 2]
+            UBRI = UB[:n // 2, n // 2:]
+            UBII = UB[n // 2:, n // 2:]
+            uRR = _prod(RA, _prod(UBRR, RA)) - _prod(IA,
+                                                     _prod(UBRI.T, RA)) - _prod(
+                RA, _prod(UBRI, IA)) + _prod(IA, _prod(UBII, IA))
+            uRI = _prod(RA, _prod(UBRR, IA)) - _prod(IA,
+                                                     _prod(UBRI.T, IA)) + _prod(
+                RA, _prod(UBRI, RA)) - _prod(IA, _prod(UBII, RA))
+            uII = _prod(IA, _prod(UBRR, IA)) + _prod(RA,
+                                                     _prod(UBRI.T, IA)) + _prod(
+                IA, _prod(UBRI, RA)) + _prod(RA, _prod(UBII, RA))
+        return uRR, uRI, uII
+
+    N = len(Y)
+    RY = Y[:N // 2]
+    IY = Y[N // 2:]  # decompose into block matrix
+    RF = F[:N // 2]
+    IF = F[N // 2:]  # decompose into block matrix
+    YF = np.r_[RY * RF - IY * IF, RY * IF + IY * RF]  # apply product rule
+    if not isinstance(UF, np.ndarray):  # second factor is known exactly
+        UYRR, UYRI, UYII = calcU(F, UY)
+        UYF = np.vstack((np.hstack((UYRR, UYRI)), np.hstack(
+            (UYRI.T, UYII))))  # stack together covariance matrix
+    else:  # both factors are uncertain
+        URR_Y, URI_Y, UII_Y = calcU(F, UY)
+        URR_F, URI_F, UII_F = calcU(Y, UF)
+        URR = URR_Y + URR_F
+        URI = URI_Y + URI_F
+        UII = UII_Y + UII_F
+        UYF = np.vstack((np.hstack((URR, URI)), np.hstack(
+            (URI.T, UII))))  # stack together covariance matrix
```

Comment:
These linebreaks do not contribute to the overall readability in my opinion. I would move the comment above the line of code.

Reply:
Agree!